design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,477 @@
+spec:
+  version: v0.0.2
+  componentRef:
+    name: frontend
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash top"
+```
+
+Now the basic `ComponentRevision` is in place, let's talk about it in Trait and ApplicationConfiguration.
+
+### ComponentRevision could be referenced by Trait 
+
+Some traits can control version of component which means they can specify a component revision and won't be affected
+by the change of component.
+
+For these traits, they have to declare `revisionAware` field to be `true` in TraitDefinition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: fancytraits.core.oam.dev
+spec:
+  revisionAware: true
+  definitionRef:
+    name: fancytraits.core.oam.dev
+```
+
+Workload instance's name emitted by OAM runtime will be decided by the `revisionAware` field.
+
+- If **NO** traits binding with a component has `revisionAware` as true, OAM runtime will let the workload itself to handle
+   versioning things. We will only have one workload instance with the same name of componentName, and update workload directly
+   when component changed.
+- If **ANY** trait binding with a component has `revisionAware` as true, OAM runtime will always create a new workload with
+   a new name which can be aligned with the name of newly created component revision. OAM runtime will leave the trait to
+   control existing workloads to do rolling update and garbage collection. 
+
+So the essential difference between true or false of `revisionAware` is workload CR's name is revisionName or componentName.
+
+For example, when we don't use any trait who has revisionAware as true, things will work like below:
+
+Firstly, we create an `ApplicationConfiguration` with definition : 
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+```
+
+The real workload instance will be created with the same name as the componentName.
+
+```yaml
+  apiVersion: core.oam.dev/v1alpha2
+  kind: ContainerizedWorkload
+  metadata:
+    name: frontend
+  spec:
+    containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+      - "bash lscpu"
+```
+
+But if we use any trait has revisionAware.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: FancyTrait
+            spec:
+              traffic:
+                - revision: frontend-v0.0.2
```

Comment:
this is not revision, it is name + revision. Shouldn't this be separate fields?

Reply:
Is `frontend-c8bb659c5` match?